pkg/aws/eni/node.go

```diff
@@ -0,0 +1,537 @@
+
+	available map[string]v2.AllocationIP
+
+	manager *NodeManager
+}
+
+type nodeStatistics struct {
+	// usedIPs is the number of IPs currently in use
+	usedIPs int
+
+	// availableIPs is the number of IPs currently available for allocation
+	// by the node
+	availableIPs int
+
+	// neededIPs is the number of IPs needed to reach the PreAllocate
+	// watermwark
+	neededIPs int
+
+	// remainingInterfaces is the number of ENIs that can either be
+	// allocated or have not yet exhausted the ENI specific quota of
+	// addresses
+	remainingInterfaces int
+}
+
+func (n *Node) logger() *logrus.Entry {
+	if n == nil {
+		return log
+	}
+
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	return n.loggerLocked()
+}
+
+func (n *Node) loggerLocked() *logrus.Entry {
+	if n == nil {
+		return log
+	}
+
+	logger := log.WithField(fieldName, n.name)
+
+	if n.resource != nil {
+		logger = logger.WithField("instanceID", n.resource.Spec.ENI.InstanceID)
+	}
+
+	return logger
+}
+
+func (n *Node) getNeededAddresses() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	return n.stats.neededIPs
+}
+
+func calculateNeededIPs(availableIPs, usedIPs, preAllocate, minAllocate int) (neededIPs int) {
+	if preAllocate == 0 {
+		preAllocate = defaults.ENIPreAllocation
+	}
+
+	neededIPs = preAllocate - (availableIPs - usedIPs)
+	if neededIPs < 0 {
+		neededIPs = 0
+	}
+
+	if minAllocate > 0 {
+		neededIPs = math.IntMax(neededIPs, minAllocate-availableIPs)
+	}
+
+	return
+}
+
+func (n *Node) updatedResource(resource *v2.CiliumNode) bool {
+	n.mutex.Lock()
+	n.resource = resource
+	allocationNeeded := n.recalculateLocked()
+	n.mutex.Unlock()
+
+	if allocationNeeded {
+		n.manager.deficitResolver.TriggerWithReason(resource.Name)
+	}
+
+	return allocationNeeded
+}
+
+func (n *Node) recalculateLocked() bool {
+	n.enis = map[string]v2.ENI{}
+	n.available = map[string]v2.AllocationIP{}
+	for _, e := range n.manager.instancesAPI.GetENIs(n.resource.Spec.ENI.InstanceID) {
+		n.enis[e.ID] = *e
+
+		if e.Number < n.resource.Spec.ENI.FirstInterfaceIndex {
+			continue
+		}
+
+		for _, ip := range e.Addresses {
+			n.available[ip] = v2.AllocationIP{Resource: e.ID}
+		}
+	}
+	n.stats.usedIPs = len(n.resource.Status.IPAM.InUse)
+	n.stats.availableIPs = len(n.available)
+	n.stats.neededIPs = calculateNeededIPs(n.stats.availableIPs, n.stats.usedIPs, n.resource.Spec.ENI.PreAllocate, n.resource.Spec.ENI.MinAllocate)
+	allocationNeeded := !n.resyncNeeded && !n.waitingForAllocation && n.stats.neededIPs > 0
+
+	n.loggerLocked().WithFields(logrus.Fields{
+		"available":            n.stats.availableIPs,
+		"used":                 n.stats.usedIPs,
+		"toAlloc":              n.stats.neededIPs,
+		"resyncNeeded":         n.resyncNeeded,
+		"waitingForAllocation": n.waitingForAllocation,
+	}).Info("Recalculated needed addresses")
+
+	if allocationNeeded {
+		n.waitingForAllocation = true
+		n.resyncNeeded = true
+	}
+
+	return allocationNeeded
+}
+
+func (n *Node) resourceCopy() *v2.CiliumNode {
```

Comment:
I don't understand this comment

Reply:
@tgraf `resourceCopy() ` is a private method but it's also doing a RLock. Consider making the function public so that it is clear it can be accessed without worrying about the Lock being held or not or add a comment in the function saying that RLock is not required to be held.

Reply:
Ah, now I get it. Done.